### PR TITLE
Implementing build annotations.

### PR DESCRIPTION
**Pull request by cprov**
Recording `reason` (text) why builds are triggered (`/launchpad/snaps/request-builds`) and returning corresponding 'build_annotations' map when builds are retrieved (`/launchpad/builds`)

### CODE REVIEW

**Comment by bartaz** [comment]
Quite recently we had issues with relying on human readable messages (#979), so I would feel safer if we didn't repeat the same with build annotations.

It's quite likely that UI will display these annotations differently from the messages stored in the database, so they will need to be mapped into other text, icon or whatever UI chooses to use.
It would be better if database would keep identifiers less prone to change (codes like "BUILD_REASON_MANUAL" for example).

Or at least it would be good to define all possible reasons (used currently in the app) in one place and only define these human readable text representation there. In all other places we should reference these messages by variable name to avoid changing them in many places.

**Comment by cprov** [comment]
@bartaz thanks for the direction, I've updated it to store identifiers.

This implementation is not quite what I envisioned for "Buildsets" (group of snap revisions with a common trigger and possibly a changelog message), but it would drastically affect the UX. Plain annotations is less smart but suits the existing UX better.

Let me know if we need more fixes before landing it.

File: src/server/handlers/launchpad.js [hunk 1]
```diff
@@ -672,19 +672,32 @@ export const getSnapBuilds = async (req, res) => {
     const snap = await getLaunchpad().get(snapUrl);
     const builds = await internalGetSnapBuilds(snap, req.query.start, req.query.size);
 
+    // Let's collect corresponding build annotations.
+    let build_annotations = {};
```
Comment (bartaz) [inline src/server/handlers/launchpad.js]: Ideally it would be great for collecting annotations to be part of `internalGetSnapBuilds`, so that `internalGetSnapBuilds` already returns annotated builds. In current implementation annotations are returned separately to the UI, which means that it's the front-end that will need to join annotations with build object.
Preferably this join would be done in `internapGetSnapBuilds`, so that any part of the app that uses `internalSnapBuilds` would get annotated builds.

If that seems unnecessary, I'd suggest at least extracting getting annotations to some `internalGetBuildAnnotations` method, so the logic is not hidden only in `getSnapBuilds` handler.

Comment (cprov) [reply src/server/handlers/launchpad.js]: I've isolated `internalGetBuildAnnotations` and `getBuildId(<build>)` common helper so the client work is clearer. I thought it's better to keep build records pristine, as they come from LP, for the time being.

File: src/server/handlers/launchpad.js [hunk 2]
```diff
@@ -726,8 +760,9 @@ export const requestSnapBuilds = async (req, res) => {
     const { owner, name } = await checkAdminPermissions(
       req.session, req.body.repository_url
     );
+    const reason = req.body.reason || 'triggered-manually';
```
[the remark this answers is not in the document]
Comment (cprov) [reply src/server/handlers/launchpad.js]: Right, it's definitely cleaner now. Thanks for the suggestion.

File: src/server/handlers/launchpad.js [hunk 3]
```diff
@@ -726,8 +769,9 @@ export const requestSnapBuilds = async (req, res) => {
     const { owner, name } = await checkAdminPermissions(
       req.session, req.body.repository_url
     );
+    const reason = req.body.reason || BUILD_TRIGGERED_MANUALLY;
     const snap = await internalFindSnap(req.body.repository_url);
-    const builds = await internalRequestSnapBuilds(snap, owner, name);
+    const builds = await internalRequestSnapBuilds(snap, owner, name, reason);
     return res.status(201).send({
       status: 'success',
       payload: {
```
Comment (bartaz) [inline src/server/handlers/launchpad.js]: Currently `requestSnapBuilds` payload only contains builds.
It would be good if it would return `build_annotations` as well, similar to `getSnapBuilds`, so once new builds are requested we can immediately show them in UI with annotations (without refetching them from server).

File: test/routes/src/server/routes/t_webhook.js [hunk 4]
```diff
@@ -307,20 +307,16 @@ describe('The WebHook API endpoint', () => {
...
-                    self_link: `${lp_api_base}${lp_snap_path}/+build/1`
-                  },
-                  {
-                    resource_type_link: `${lp_api_base}/#snap_build`,
-                    self_link: `${lp_api_base}${lp_snap_path}/+build/2`
-                  }
-                ]
-              });
+              .reply(200, [
+                {
```
Comment (bartaz) [inline test/routes/src/server/routes/t_webhook.js]: Why is this change added?
Doesn't LP collection response contain `total_size` etc properties?

Comment (cprov) [reply test/routes/src/server/routes/t_webhook.js]: Not the response of `requestAutoBuilds`, it simply returns the created builds in the body. It must have been confused with the response of snap builds which is indeed a LP collection with total_size and start.

File: src/server/handlers/launchpad.js (hunk 3 above)
Comment (cprov) [reply src/server/handlers/launchpad.js]: Good point, I will include the just created build_annotations.
